### PR TITLE
Use cabal new-build for testing

### LINKED ISSUE #225: Figure out a way to cache the Travis build
Currently, the Travis CI job for `singletons` rebuilds all of its dependencies, every time. This makes me immensely sad.

Unfortunately, actually making Travis cache the dependencies is easier said than done. The thorn in our side is the need to install `th-desugar` from a submodule separately from all other dependencies. The usual Travis script that caches Hackage dependencies, however, calculates whether dependencies have changed or not with this command:

```bash
$ cabal install --only-dependencies --enable-tests --enable-benchmarks --dry -v > installplan.txt
```

You can't just throw `th-desugar` in there, since it doesn't play nicely with `--only-dependencies` and `--enable-tests`.

The proper solution would be to switch over to a `.travis.yml` file that uses `cabal new-build` (i.e., the approach [here](https://github.com/hvr/multi-ghc-travis/blob/800980d76f7a74f3cdfd76b3dff351d52d2c84ee/.travis.yml)). But migrating over to `cabal new-build` comes with its own set of problems, as the `singletons` test suite is designed with some "old-build" assumptions in mind. Namely,

1. It assumes that a file named `"../dist/build/autogen/cabal_macros.h"` exists. This assumption doesn't hold under `cabal new-build`.
2. It assumes that you're either using the global package database or one from a sandbox. `cabal new-build` organizes its own package databases, however.

@hvr had some ideas for porting the test suite over. Some possibilities:

1. Use a `Setup.hs` script to determine the right paths for `cabal_macros.h`/package databases. (I'm not especially eager to do this, since `Setup.hs` scripts come with their own problems.)
2. Augment `Paths_singletons` with the necessary information. This would require waiting for some number of `Cabal` releases, but it would likely be a clean solution.

### PR DESCRIPTION
**Pull request by RyanGlScott**
This finally allows us to test `singletons` on GHC 8.4.1 and HEAD. The new way to invoke tests is now:

```
# Assuming you have ghc-8.4.1 and cabal-install-2.1 or later on your PATH
cabal new-configure --enable-tests
cabal new-build
cabal new-test # Assuming you are in the directory containing singletons.cabal
```

This changes the test suite to invoke `cabal new-exec ghc -- ...`, which piggybacks on the previously built files in the store. A bit fiddly, but it does work.

Fixes #225.

### CODE REVIEW

**Comment by RyanGlScott** [comment]
@goldfirere, this approach does require the use of a development version of `cabal-install` (slated to be released sometime alongside GHC 8.4), so I'm not sure if we should foist that requirement onto everyone running the tests. What do you think?

File: tests/compile-and-dump/Promote/GenDefunSymbols.ghc84.template [hunk 1]
```diff
@@ -9,7 +9,7 @@ Promote/GenDefunSymbols.hs:0:0:: Splicing declarations
         = snd ((GHC.Tuple.(,) LiftMaybeSym1KindInference) GHC.Tuple.())
     data LiftMaybeSym1 (l :: TyFun a0123456789876543210 b0123456789876543210
                              -> Type) (l :: TyFun (Maybe a0123456789876543210) (Maybe b0123456789876543210))
-      = forall arg. Data.Singletons.Internal.SameKind (Apply (LiftMaybeSym1 l) arg) (LiftMaybeSym2 l arg) =>
+      = forall arg. singletons-2.4.1:Data.Singletons.Internal.SameKind (Apply (LiftMaybeSym1 l) arg) (LiftMaybeSym2 l arg) =>
```
Comment (goldfirere) [inline tests/compile-and-dump/Promote/GenDefunSymbols.ghc84.template]: Can we scrub the version number from these files before comparing? That seems more robust.

**Comment by RyanGlScott** [comment]
For me, the most pressing issue is just being able to have tests and Haddocks successfully running on each commit. It's nice in theory to say "we'll just test before pushing", but sometimes we forget to check! (Plus, if other folks make pull requests, it'd be nice to check their work too.)

I'll let you make the final call on whether to merge now or later.

File: tests/compile-and-dump/Promote/GenDefunSymbols.ghc84.template (hunk 1 above)
Comment (RyanGlScott) [reply tests/compile-and-dump/Promote/GenDefunSymbols.ghc84.template]: Oh dear, I somehow missed that. Yes, we should absolutely scrub versions out using the `sed` machinery.

**Comment by goldfirere** [comment]
Well, I suppose the package only works with a pre-release GHC, so I suppose it's OK if it also only works with a pre-release cabal. Go for it.

**Pull request merged**